File: c++/memkvstore_test.cc
#include <gtest/gtest.h>

#include "memkvstore.h"

std::string const key1 = "foo";
std::string const val1 = "bar";

std::string const key2 = "baz";
std::string const val2 = "quux";

using multipaxos::Command;
using multipaxos::CommandType;
using multipaxos::CommandType::DEL;
using multipaxos::CommandType::GET;
using multipaxos::CommandType::PUT;

TEST(MemKVStoreTest, GetPutDel) {
  kvstore::MemKVStore store;

  EXPECT_EQ(std::nullopt, store.Get(key1));

  EXPECT_FALSE(store.Del(key1));

  EXPECT_TRUE(store.Put(key1, val1));
  EXPECT_EQ(val1, *store.Get(key1));

  EXPECT_TRUE(store.Put(key2, val2));
  EXPECT_EQ(val2, *store.Get(key2));

  EXPECT_TRUE(store.Put(key1, val2));
  EXPECT_EQ(val2, *store.Get(key1));
  EXPECT_EQ(val2, *store.Get(key2));

  EXPECT_TRUE(store.Del(key1));
  EXPECT_EQ(std::nullopt, store.Get(key1));
  EXPECT_EQ(val2, *store.Get(key2));

  EXPECT_TRUE(store.Del(key2));
  EXPECT_EQ(std::nullopt, store.Get(key1));
  EXPECT_EQ(std::nullopt, store.Get(key2));
}

Command MakeCommand(CommandType type,
                    std::string const& key,
                    std::string const& value) {
  Command cmd;
  cmd.set_type(type);
  cmd.set_key(key);
  cmd.set_value(value);
  return cmd;
}

TEST(MemKVStoreTest, Execute) {
  auto get_key1 = MakeCommand(GET, key1, "");
  auto get_key2 = MakeCommand(GET, key2, "");
  auto del_key1 = MakeCommand(DEL, key1, "");
  auto put_key1val1 = MakeCommand(PUT, key1, val1);
  auto put_key2val2 = MakeCommand(PUT, key2, val2);
  auto put_key1val2 = MakeCommand(PUT, key1, val2);

  kvstore::MemKVStore store;
  {
    kvstore::KVResult r = kvstore::Execute(get_key1, &store);
    EXPECT_TRUE(!r.ok_ && r.value_ == kvstore::kNotFound);
  }
  {
    kvstore::KVResult r = kvstore::Execute(del_key1, &store);
    EXPECT_TRUE(!r.ok_ && r.value_ == kvstore::kNotFound);
  }
  {
    kvstore::KVResult r1 = kvstore::Execute(put_key1val1, &store);
    EXPECT_TRUE(r1.ok_ && r1.value_ == kvstore::kEmpty);

    kvstore::KVResult r2 = kvstore::Execute(get_key1, &store);
    EXPECT_TRUE(r2.ok_ && r2.value_ == val1);
  }
  {
    kvstore::KVResult r1 = kvstore::Execute(put_key2val2, &store);
    EXPECT_TRUE(r1.ok_ && r1.value_ == kvstore::kEmpty);

    kvstore::KVResult r2 = kvstore::Execute(get_key2, &store);
    EXPECT_TRUE(r2.ok_ && r2.value_ == val2);
  }
  {
    kvstore::KVResult r1 = kvstore::Execute(put_key1val2, &store);
    EXPECT_TRUE(r1.ok_ && r1.value_ == kvstore::kEmpty);

    kvstore::KVResult r2 = kvstore::Execute(get_key1, &store);
    EXPECT_TRUE(r2.ok_ && r2.value_ == val2);

    kvstore::KVResult r3 = kvstore::Execute(get_key2, &store);
    EXPECT_TRUE(r3.ok_ && r3.value_ == val2);
  }
  {
    kvstore::KVResult r1 = kvstore::Execute(del_key1, &store);
    EXPECT_TRUE(r1.ok_ && r1.value_ == kvstore::kEmpty);

    kvstore::KVResult r2 = kvstore::Execute(get_key1, &store);
    EXPECT_FALSE(r2.ok_ && r2.value_ == kvstore::kNotFound);

    kvstore::KVResult r3 = kvstore::Execute(get_key2, &store);
    EXPECT_TRUE(r3.ok_ && r3.value_ == val2);
  }
}

File: go/kvstore/memkvstore_test.go
package kvstore

import (
	pb "github.com/psu-csl/replicated-store/go/multipaxos/network"
	"github.com/stretchr/testify/assert"
	"testing"
)

const (
	key1 string = "foo"
	val1 string = "bar"
	key2 string = "baz"
	val2 string = "qux"
)

func TestMemKVStore_GetPutDel(t *testing.T) {
	store := NewMemKVStore()

	assert.Nil(t, store.Get(key1))

	assert.False(t, store.Del(key1))

	assert.True(t, store.Put(key1, val1))
	actualVal := *store.Get(key1)
	assert.Equal(t, val1, actualVal)

	assert.True(t, store.Put(key2, val2))
	actualVal = *store.Get(key2)
	assert.Equal(t, val2, actualVal)

	assert.True(t, store.Put(key1, val2))
	actualVal = *store.Get(key1)
	assert.Equal(t, val2, actualVal)
	actualVal = *store.Get(key2)
	assert.Equal(t, val2, actualVal)

	assert.True(t, store.Del(key1))
	assert.Nil(t, store.Get(key1))
	actualVal = *store.Get(key2)
	assert.Equal(t, val2, actualVal)

	assert.True(t, store.Del(key2))
	assert.Nil(t, store.Get(key1))
	assert.Nil(t, store.Get(key2))
}

func TestMemKVStore_Execute(t *testing.T) {
	store := NewMemKVStore()
	getKey1 := &pb.Command{Key: key1, Value: "", Type: pb.Get}
	getKey2 := &pb.Command{Key: key2, Value: "", Type: pb.Get}
	delKey1 := &pb.Command{Key: key1, Value: "", Type: pb.Del}
	putKey1Val1 := &pb.Command{Key: key1, Value: val1, Type: pb.Put}
	putKey2Val2 := &pb.Command{Key: key2, Value: val2, Type: pb.Put}
	putKey1Val2 := &pb.Command{Key: key1, Value: val2, Type: pb.Put}

	{
		r := Execute(getKey1, store)
		assert.True(t, !r.Ok && r.Value == NotFound)
	}

	{
		r := Execute(delKey1, store)
		assert.True(t, !r.Ok && r.Value == NotFound)
	}

	{
		r1 := Execute(putKey1Val1, store)
		assert.True(t, r1.Ok && r1.Value == Empty)
		r2 := Execute(getKey1, store)
		assert.True(t, r2.Ok && r2.Value == val1)
	}

	{
		r1 := Execute(putKey2Val2, store)
		assert.True(t, r1.Ok && r1.Value == Empty)
		r2 := Execute(getKey2, store)
		assert.True(t, r2.Ok && r2.Value == val2)
	}

	{
		r1 := Execute(putKey1Val2, store)
		assert.True(t, r1.Ok && r1.Value == Empty)

		r2 := Execute(getKey1, store)
		assert.True(t, r2.Ok && r2.Value == val2)

		r3 := Execute(getKey2, store)
		assert.True(t, r3.Ok && r3.Value == val2)
	}

	{
		r1 := Execute(delKey1, store)
		assert.True(t, r1.Ok && r1.Value == Empty)

		r2 := Execute(getKey1, store)
		assert.False(t, r2.Ok)
		assert.Equal(t, NotFound, r2.Value)

		r3 := Execute(getKey2, store)
		assert.True(t, r3.Ok && r3.Value == val2)
	}
}
